File: 백준/Gold/17472. 다리 만들기 2/다리 만들기 2.py
from collections import deque
from heapq import heappush, heappop

n, m = map(int, input().split())

board = [list(map(int, input().split())) for _ in range(n)]
graph = []


def bfs(y, x, _visit, _number):
    global n, m, board
    
    que = deque()
    que.append((y, x))
    _visit[y][x] = True
    board[y][x] = _number
    
    dy, dx = (0,0,1,-1), (1,-1,0,0)

    while que:
        cur_y, cur_x = que.popleft()

        for i in range(4):
            next_y, next_x = cur_y + dy[i], cur_x + dx[i]

            if 0 > next_y or next_y >= n or 0 > next_x or next_x >= m:
                continue

            if _visit[next_y][next_x] or board[next_y][next_x] == 0:
                continue

            que.append((next_y, next_x))
            _visit[next_y][next_x] = True
            board[next_y][next_x] = _number


def connect_horizon(y, x):
    global n, m, board, graph
    
    dist = 0

    sour = board[y][x]
    for i in range(x - 1, -1, -1):
        des = board[y][i]
        
        if des == sour:
            break

        if des == 0:
            dist += 1
            continue

        if dist == 1:
            break
        
        heappush(graph, (dist, sour, des))
        break
        
    dist = 0
    for i in range(x + 1, m):
        des = board[y][i]
        if des == sour:
            break

        if des == 0:
            dist += 1
            continue

        if dist == 1:
            break
        
        heappush(graph, (dist, sour, des))
        break


def connect_vertical(y, x):
    global n, m, board, graph

    dist = 0

    sour = board[y][x]
    for i in range(y - 1, -1, -1):
        des = board[i][x]
        
        if des == sour:
            break

        if des == 0:
            dist += 1
            continue

        if dist == 1:
            break
        
        heappush(graph, (dist, sour, des))
        break
        
    dist = 0
    for i in range(y + 1, n):
        des = board[i][x]
        if des == sour:
            break

        if des == 0:
            dist += 1
            continue

        if dist == 1:
            break
        
        heappush(graph, (dist, sour, des))
        break


def find_parent(x):
    global parents

    if parents[x] == x:
        return x

    parents[x] = find_parent(parents[x])

    return parents[x]


def union(x, y):
    global parents

    px, py = find_parent(x), find_parent(y)

    if px == py:
        return

    parents[py] = px


number = 1
visit = [[False] * m for _ in range(n)]

# 섬 번호 붙이기
for i in range(n):
    for j in range(m):
        if board[i][j] and not visit[i][j]:
            bfs(i, j, visit, number)
            number += 1


# 섬 연결
for i in range(n):
    for j in range(m):
        if board[i][j]:
            connect_vertical(i, j)
            connect_horizon(i, j)

#MST
answer = 0
parents = [i for i in range(number)]
visit = [[False] * number for _ in range(number)]

while graph:
    cost, sour, des = heappop(graph)
    if visit[sour][des]:
        continue
    
    sour_parent, des_parent = find_parent(sour), find_parent(des)

    if sour_parent == des_parent:
        continue

    union(sour_parent, des_parent)
    visit[sour][des] = visit[des][sour] = True
    answer += cost
    

# 다 같은 집합에 존재하는 확인
parent = find_parent(1)
flag = True
for child in range(2, number):
    if parent != find_parent(child):
        flag = False
        break

    
print(answer if flag else -1)

File: 백준/Gold/17472. 다리 만들기 2/다리 만들기 2.java
import java.io.*;
import java.util.*;

class Main{
    private static int n, m;
    private static int[][] map;
    private static int[][] islands;
    private static PriorityQueue<int[]> pq;
    private static boolean[][] visit;
    private static int[] dy = {0,0,1,-1}, dx = {1,-1,0,0}, parents;

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        StringTokenizer tokenizer = new StringTokenizer(br.readLine());

        n = Integer.parseInt(tokenizer.nextToken());
        m = Integer.parseInt(tokenizer.nextToken());

        map = new int[n][m];
        for(int i=0 ; i<n ; i++){
            tokenizer = new StringTokenizer(br.readLine());

            for(int j=0 ; j<m ; j++){
                map[i][j] = Integer.parseInt(tokenizer.nextToken());
            }
        }

        int number = 1;
        visit = new boolean[n][m];
        islands = new int[n][m];
        for(int i=0 ; i<n ; i++){
            for(int j=0 ; j<m ; j++){
                if(visit[i][j] || map[i][j] == 0)
                    continue;

                numbering(number, i, j);
                number++;
            }
        }

        pq = new PriorityQueue<>((o1, o2) -> o1[2] - o2[2]);
        parents = new int[number];
        for(int i=0 ; i<number; i++)
            parents[i] = i;

        for(int i=0 ; i<n ; i++){
            for(int j=0 ; j<m ; j++){
                if(islands[i][j] == 0)
                    continue;

                connecting(i, j, islands[i][j]);
            }
        }

        int answer = MST();
        int px = findParent(1);
        for(int i=2 ; i<number ; i++){
            if(px != findParent(i)){
                answer = -1;
                break;
            }
        }

        bw.write(String.valueOf(answer));
		bw.close();
		br.close();
	}

    private static int findParent(int x){
        if(x == parents[x])
            return x;
        return parents[x] = findParent(parents[x]);
    }

    private static void union(int x, int y){
        int px = findParent(x), py = findParent(y);

        if(px == py)
            return;
        
        parents[py] = px;
    }

    private static int MST(){
        int ret = 0;

        while(!pq.isEmpty()){
            int[] cur = pq.poll();

            int sour = cur[0], des = cur[1], dist = cur[2];
            if(findParent(sour) == findParent(des))
                continue;
                
            union(sour, des);
            ret += dist;
        }

        return ret;
    }

    private static boolean isIn(int y, int x){
        return 0 <= y && y < n && 0 <= x && x < m;
    }

    private static void connecting(int y, int x, int number){
        for(int i=0 ; i<4 ; i++){
            for(int j=1 ; j<Math.max(n, m) ; j++){
                int ny = y + dy[i] * j, nx = x + dx[i] * j;
                
                if(!isIn(ny, nx) || islands[ny][nx] == number)
                    break;
                
                int idx = islands[ny][nx];
                if(idx == 0)
                    continue;
                
                if(j - 1 <= 1)
                    break;

                pq.add(new int[]{number, idx, j - 1});
                break;
            }
        }
    }

    
    private static void numbering(int number, int startY, int startX){
        Queue<int[]> que = new ArrayDeque<>();
        que.add(new int[]{startY, startX});

        visit[startY][startX] = true;

        while(!que.isEmpty()){
            int[] cur = que.poll();

            int y = cur[0], x = cur[1];
            islands[y][x] = number;

            for(int i=0 ; i<4 ; i++){
                int ny = y + dy[i], nx = x + dx[i];

                if(!isIn(ny, nx) || visit[ny][nx] || map[ny][nx] == 0)
                    continue;

                visit[ny][nx] = true;
                que.add(new int[]{ny, nx});
            }
        }
    }

}